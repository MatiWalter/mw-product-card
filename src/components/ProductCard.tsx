import React from 'react';
import { useProduct } from '../hooks/useProduct';
import { 
  InitialValues,
  onChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from '../interfaces';

import styles from '../styles/styles.module.css';

export const ProductContext = React.createContext<ProductContextProps>({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  className?: string;
  initialValues?: InitialValues;
  product: Product;
  style?: React.CSSProperties;
  value?: number;
  children: ( args: ProductCardHandlers ) => JSX.Element;
  onChange?: (args: onChangeArgs) => void;
}

export const ProductCard = ({ 
  children,
  className,
  initialValues,
  onChange,
  product,
  style,
  value
}: Props) => {

  const { 
    counter,
    isMaxCountReached,
    maxCount,
    increaseBy,
    reset
  } = useProduct({ initialValues, onChange, product, value });

  return (
    <Provider value={{ counter, increaseBy, product, maxCount }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        { 
          children({
            count: counter,
            isMaxCountReached,
            maxCount: initialValues?.maxCount,
            product,
            increaseBy,
            reset
          })
        }
      </div>
    </Provider>
  )
}
