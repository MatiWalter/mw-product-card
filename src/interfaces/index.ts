export type { 
  InitialValues,
  onChangeArgs,
  Product,
  ProductCardHandlers,
  ProductCardHOCProps,
  ProductContextProps,
  ProductInCart,
} from './interfaces';