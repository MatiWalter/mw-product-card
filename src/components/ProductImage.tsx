import React from 'react';
import { useContext } from 'react';
import { ProductContext } from './ProductCard';

import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export interface Props {
  img?: string;
  className?: string;
  style?: React.CSSProperties 
}

export const ProductImage = ({ img, className, style }: Props) => {
  
  const { product } = useContext(ProductContext);

  let image = 
    img 
      ? img 
      : product.img 
        ? product.img 
        : noImage;

  return (
    <img className={`${styles.productImg} ${className}`} style={style} src={image} alt="Product Img" />
  )
}