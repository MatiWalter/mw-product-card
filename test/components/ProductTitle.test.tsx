import React from 'react'
import renderer from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../data/products';

describe('ProductTitle', () => {
  test('should render component with custom title', () => {
    const wrapper = renderer.create(
      <ProductTitle title="Custom Product" className='custom-class' />
    );

    expect(wrapper).toMatchSnapshot();
  });

  test('should render default title if no prop is passed', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {
          () => (
            <ProductTitle />
          )
        }
      </ProductCard> 
    );

    expect(wrapper).toMatchSnapshot();
  });
});
