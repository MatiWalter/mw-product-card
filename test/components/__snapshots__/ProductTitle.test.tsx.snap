// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle should render component with custom title 1`] = `
<span
  className="productDescription custom-class"
>
  Custom Product
</span>
`;

exports[`ProductTitle should render default title if no prop is passed 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug - Sin imagen
  </span>
</div>
`;
